syntax = "proto3";

package chat;

message TextMessage {
  string sender = 1;
  string content = 2;
  string target = 3;
  TargetType target_type = 4;
}

enum TargetType {
  USER = 0;
  GROUP = 1;
}

message TextMessageResponse {
  bool success = 1;
  string error_message = 2;
}

message LoginRequest {
  string username = 1;
  string password = 2;
}

message LoginResponse {
  string username = 1;
  bool success = 2;
  string error_message = 3;
}

message LogoutRequest {
  string username = 1;
}

message LogoutResponse {
  bool success = 1;
  string error_message = 2;
}

message RegisterRequest {
  string username = 1;
  string password = 2;
}

message RegisterResponse {
  bool success = 1;
  string error_message = 2;
}

message SearchRequest {
  string keyword = 1;
  bool online_only = 2;
}

message SearchResponse {
  repeated string usernames = 1;
}

enum GroupOperation {
  CREATE = 0;
  JOIN = 1;
  LEAVE = 2;
  QUERY = 3;
}

message GroupRequest {
  GroupOperation operation = 1;
  string group_name = 2;
  string username = 3;
}

message GroupResponse {
  GroupOperation operation = 1;
  bool success = 2;
  string error_message = 3;
  repeated string joined_groups = 4;
}